File: src/main/java/ru/citeck/ecos/uiserv/domain/journal/service/JournalService.java
package ru.citeck.ecos.uiserv.domain.journal.service;

import ru.citeck.ecos.records2.predicate.model.Predicate;
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalDef;
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalWithMeta;

import java.util.List;
import java.util.Set;
import java.util.function.BiConsumer;

public interface JournalService {

    long getLastModifiedTimeMs();

    JournalWithMeta getJournalById(String id);

    List<JournalWithMeta> getAll(int max, int skipCount, Predicate predicate);

    JournalWithMeta getById(String id);

    Set<JournalWithMeta> getAll(int maxItems, int skipCount);

    Set<JournalWithMeta> getAll(Set<String> extIds);

    long getCount();

    long getCount(Predicate predicate);

    void onJournalChanged(BiConsumer<JournalDef, JournalDef> consumer);

    JournalWithMeta save(JournalDef dto);

    void delete(String id);
}

File: src/main/java/ru/citeck/ecos/uiserv/domain/journal/service/JournalService.kt
package ru.citeck.ecos.uiserv.domain.journal.service

import ru.citeck.ecos.model.lib.workspace.IdInWs
import ru.citeck.ecos.records2.predicate.model.Predicate
import ru.citeck.ecos.records3.record.dao.query.dto.query.SortBy
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalDef
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalWithMeta
import ru.citeck.ecos.uiserv.domain.journal.service.provider.JournalsProvider
import java.util.function.BiConsumer
import java.util.function.Consumer

interface JournalService {

    fun getLastModifiedTimeMs(): Long

    fun getJournalById(id: IdInWs): JournalWithMeta?

    fun getAll(predicate: Predicate, max: Int, skip: Int, sort: List<SortBy>): List<JournalWithMeta>

    fun getAll(maxItems: Int, skipCount: Int): Set<JournalWithMeta>

    fun getAll(extIds: Collection<IdInWs>): List<JournalWithMeta>

    fun getCount(): Long

    fun getCount(predicate: Predicate): Long

    fun onJournalDeleted(consumer: Consumer<JournalWithMeta>)

    fun onJournalChanged(consumer: BiConsumer<JournalDef?, JournalDef?>)

    fun onJournalWithMetaChanged(consumer: BiConsumer<JournalWithMeta?, JournalWithMeta?>)

    fun save(dto: JournalDef): JournalWithMeta

    fun delete(id: IdInWs)

    fun registerProvider(provider: JournalsProvider)
}
